File: SearchInsertDelete/SearchInsertDelete/sidlist.cpp
#include "sidlist.h"

#include <iostream>
#include <algorithm>
#include <thread>

using namespace std;

SIDList::SIDList(std::vector<int>& v)
    : m_list(move(v))
{

}

int SIDList::search(int value)
{
    {
        unique_lock<mutex> guard(m_mt);
        m_cv.wait(guard, [&]() {
            return m_numConcurrentDels==0;
        });
        m_numConcurrentSearches++;
        this->print_status();
    }

    this_thread::sleep_for(chrono::milliseconds(500));
    auto it = find(m_list.begin(), m_list.end(), value);
    m_numConcurrentSearches--;
    m_cv.notify_all();
    if(it != m_list.end()) {
        return -1;
    }
    return static_cast<int>(distance(m_list.begin(), it));
}

void SIDList::insert(int value)
{
    {
        unique_lock<mutex> guard(m_mt);
        m_cv.wait(guard, [&]() {
            return m_numConcurrentDels==0 && m_numConcurrentInserts==0;
        });
        m_numConcurrentInserts++;
        this->print_status();
    }
    this_thread::sleep_for(chrono::milliseconds(500));
    m_list.push_back(value);
    m_numConcurrentInserts--;
    m_cv.notify_all();
}

bool SIDList::del(int indx)
{
    if(indx >= static_cast<int>(m_list.size())) {
        return false;
    }
    {
        unique_lock<mutex> guard(m_mt);
        m_cv.wait(guard, [&](){
            return m_numConcurrentDels==0 && m_numConcurrentInserts==0 && m_numConcurrentSearches==0;
        });
        m_numConcurrentDels++;
        this->print_status();
    }

    m_list.erase(next(m_list.begin(), indx));
    m_numConcurrentDels--;
    m_cv.notify_all();
    return true;
}

bool SIDList::empty()
{
    lock_guard<mutex> guard(m_mt);
    return m_list.empty();
}

void SIDList::print_status()
{
    cout << "Searches: " << m_numConcurrentSearches
         << ", Inserts: " << m_numConcurrentInserts
         << ", Deletes: " << m_numConcurrentDels;

    if(m_numConcurrentDels>0 && (m_numConcurrentSearches || m_numConcurrentSearches)) {
        cout << endl << "--Error Caught--" << endl;
    }

    if(m_numConcurrentInserts > 1) {
        cout << endl << "--Error Caught--" << endl;
    }


    if(m_numConcurrentSearches > 1 && m_numConcurrentInserts == 1) {
        cout << endl << "**We are good (just to make sure things like this happens)**" << endl;
    }

    if(m_numConcurrentDels == 1) {
        cout << endl << "**We are good (just to make sure things like this happens)**" << endl;
    }


    cout << "\r" << flush;
}

File: SearchInsertDelete/SearchInsertDelete-Go/sidlist.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

var numConcurrentSearches uint64 = 0
var numConcurrentInsertes uint64 = 0
var numConcurrentDels uint64 = 0

var mutex = sync.Mutex{}
var cond = sync.NewCond(&mutex)

type SIDList struct {
	list []int
}

func makeSIDList(list []int) SIDList {
	return SIDList{list}
}

func (e SIDList) search(value int, wg *sync.WaitGroup) int {
	if wg != nil {
		defer wg.Done()
	}
	cond.L.Lock()
	for numConcurrentDels > 0 {
		cond.Wait()
	}
	numConcurrentSearches++
	e.print_status()
	cond.L.Unlock()
	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
	for i := range e.list {
		if e.list[i] == value {
			cond.L.Lock()
			numConcurrentSearches--
			cond.L.Unlock()
			cond.Broadcast()
			return i
		}
	}
	cond.L.Lock()
	numConcurrentSearches--
	cond.L.Unlock()
	cond.Broadcast()
	return -1
}

func (e SIDList) insert(value int, wg *sync.WaitGroup) {
	if wg != nil {
		defer wg.Done()
	}
	cond.L.Lock()
	for numConcurrentDels > 0 || numConcurrentInsertes > 0 {
		cond.Wait()
	}
	numConcurrentInsertes++
	e.print_status()
	cond.L.Unlock()
	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
	e.list = append(e.list, value)
	cond.L.Lock()
	numConcurrentInsertes--
	cond.L.Unlock()
	cond.Broadcast()
}

func (e SIDList) del(idx int, wg *sync.WaitGroup) bool {
	if wg != nil {
		defer wg.Done()
	}
	if idx > len(e.list) {
		return false
	}
	cond.L.Lock()
	for numConcurrentDels > 0 ||
		numConcurrentInsertes > 0 ||
		numConcurrentSearches > 0 {
		cond.Wait()
	}
	numConcurrentDels++
	e.print_status()
	cond.L.Unlock()
	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
	//e.list = append(e.list[:idx], e.list[idx+1:]...)
	cond.L.Lock()
	numConcurrentDels--
	cond.L.Unlock()
	cond.Broadcast()
	return true
}

func (e SIDList) empty() bool {
	return len(e.list) == 0
}

func (e SIDList) print_status() {

	fmt.Print("Searches: ", numConcurrentSearches,
		", Inserts: ", numConcurrentInsertes,
		", Deletes: ", numConcurrentDels)

	if numConcurrentDels > 0 && (numConcurrentSearches > 0 || numConcurrentInsertes > 0) {

		fmt.Println("\n--Error Caught--")
	}

	if numConcurrentInsertes > 1 {
		fmt.Println("\n--Error Caught--")
	}

	if numConcurrentSearches > 1 && numConcurrentInsertes == 1 {
		fmt.Println("\n**We are good (just to make sure things like this happens)**")
	}

	if numConcurrentDels == 1 {
		fmt.Println("\n**We are good (just to make sure things like this happens)**")
	}

	fmt.Print("\r")
}
